// vuetify/lib/composables/mutationObserver.mjs
// Utilities
import { isComponentInstance } from "../util/index.mjs";
import { onBeforeUnmount, onMounted, ref, watch } from 'vue';

// Types

export function useMutationObserver(handler, options) {
  const mutationRef = ref();
  const {
    once,
    immediate,
    ...optionKeys
  } = options || {};
  const defaultValue = !Object.keys(optionKeys).length;
  const observer = new MutationObserver((mutations, observer) => {
    handler == null ? void 0 : handler(mutations, observer);
    if (options != null && options.once) observer.disconnect();
  });
  onMounted(() => {
    if (!(options != null && options.immediate)) return;
    handler == null ? void 0 : handler([], observer);
  });
  onBeforeUnmount(() => {
    observer.disconnect();
  });
  watch(mutationRef, (newValue, oldValue) => {
    if (oldValue) observer.disconnect();
    const el = isComponentInstance(newValue) ? newValue.$el : newValue;
    if (!el) return;
    observer.observe(el, {
      attributes: (options == null ? void 0 : options.attr) ?? defaultValue,
      characterData: (options == null ? void 0 : options.char) ?? defaultValue,
      childList: (options == null ? void 0 : options.child) ?? defaultValue,
      subtree: (options == null ? void 0 : options.sub) ?? defaultValue
    });
  }, {
    flush: 'post'
  });
  return {
    mutationRef
  };
}
